lib/ReactViews/Map/MapNavigation/Items/MeasureLineTool.ts

```diff
@@ -0,0 +1,158 @@
+      onCleanUp: this.onCleanUp.bind(this),
+      onMakeDialogMessage: this.onMakeDialogMessage.bind(this)
+    });
+    this.onClose = props.onClose;
+  }
+
+  get glyph(): any {
+    return GLYPHS.measure;
+  }
+
+  get viewerMode(): ViewerMode | undefined {
+    return undefined;
+  }
+
+  prettifyNumber(number: number, squared: boolean) {
+    if (number <= 0) {
+      return "";
+    }
+    // Given a number representing a number in metres, make it human readable
+    let label = "m";
+    if (number > 999) {
+      label = "km";
+      number = number / 1000.0;
+    }
+    let numberStr = number.toFixed(2);
+    // http://stackoverflow.com/questions/2901102/how-to-print-a-number-with-commas-as-thousands-separators-in-javascript
+    numberStr.replace(/\B(?=(\d{3})+(?!\d))/g, ",");
+    numberStr = `${numberStr} ${label}`;
+    return numberStr;
+  }
+
+  updateDistance(pointEntities: CustomDataSource) {
+    this.totalDistanceMetres = 0;
+    if (pointEntities.entities.values.length < 1) {
+      return;
+    }
+
+    let firstPointPos: Cartesian3 | undefined;
+    let prevPointPos: Cartesian3 | undefined;
+    for (let i = 0; i < pointEntities.entities.values.length; i++) {
+      const currentPoint = pointEntities.entities.values[i];
+      const currentPointPos = currentPoint.position!.getValue(
+        this.terria.timelineClock.currentTime
+      );
+      if (currentPointPos === undefined) continue;
+      if (prevPointPos === undefined) {
+        prevPointPos = currentPointPos;
+        firstPointPos = prevPointPos;
+        continue;
+      }
+
+      this.totalDistanceMetres =
+        this.totalDistanceMetres +
+        this.getGeodesicDistance(prevPointPos, currentPointPos);
+
+      prevPointPos = currentPointPos;
+    }
+    if (prevPointPos && firstPointPos && this.userDrawing.closeLoop) {
+      this.totalDistanceMetres =
+        this.totalDistanceMetres +
+        this.getGeodesicDistance(prevPointPos, firstPointPos);
+    }
+  }
+
+  getGeodesicDistance(pointOne: Cartesian3, pointTwo: Cartesian3) {
+    // Note that Cartesian.distance gives the straight line distance between the two points, ignoring
+    // curvature. This is not what we want.
+    const pickedPointCartographic =
+      Ellipsoid.WGS84.cartesianToCartographic(pointOne);
+    const lastPointCartographic =
+      Ellipsoid.WGS84.cartesianToCartographic(pointTwo);
+    const geodesic = new EllipsoidGeodesic(
+      pickedPointCartographic,
+      lastPointCartographic
+    );
+    return geodesic.surfaceDistance;
+  }
+
+  onCleanUp() {
+    this.totalDistanceMetres = 0;
+    super.deactivate();
+  }
+
+  onPointClicked(pointEntities: CustomDataSource) {
+    this.updateDistance(pointEntities);
+  }
+
+  onPointMoved(pointEntities: CustomDataSource) {
+    // This is no different to clicking a point.
+    this.onPointClicked(pointEntities);
+  }
+
+  onMakeDialogMessage = () => {
+    const distance = this.prettifyNumber(this.totalDistanceMetres, false);
```

Comment:
Can't this be `return this.prettifyNumber(...)` directly without the extra intermediate variables?